node/src/node/subscriptions.rs

```diff
@@ -0,0 +1,266 @@
+#[derive(Debug)]
+pub(crate) struct BroadcastingStore<S> {
+    inner: Arc<S>,
+    sender: broadcast::Sender<ExtendedHeader>,
+    last_sent_height: Option<u64>,
+    pending: Vec<Vec<ExtendedHeader>>,
+}
+
+impl<S> BroadcastingStore<S>
+where
+    S: Store,
+{
+    pub fn new(store: Arc<S>) -> Self {
+        let (sender, _) = broadcast::channel(HEIGHT_SEQUENCER_BROADCAST_CAPACITY);
+        BroadcastingStore {
+            inner: store,
+            sender,
+            last_sent_height: None,
+            pending: Default::default(),
+        }
+    }
+
+    pub fn clone_inner_store(&self) -> Arc<S> {
+        self.inner.clone()
+    }
+
+    /// Prepare BroadcastingStore for forwarding headers it received, by setting
+    /// the header height it should start at
+    pub(crate) fn init_broadcast(&mut self, head: ExtendedHeader) {
+        if self.last_sent_height.is_none() {
+            // First initialisation means Syncer has acquired a network head for the first time,
+            // start from there
+            self.last_sent_height = Some(head.height().value());
+            let _ = self.sender.send(head);
+        } else {
+            // Subsequent initialisations happen when syncer re-connects to the network
+            // this could have caused a gap in sent heights. This will get sorted out on
+            // next [`insert`].
+            self.pending.push(vec![head]);
+        }
+    }
+
+    pub(crate) fn subscribe(&self) -> broadcast::Receiver<ExtendedHeader> {
+        self.sender.subscribe()
+    }
+
+    pub(crate) async fn announce_insert(
+        &mut self,
+        range: Vec<ExtendedHeader>,
+    ) -> Result<(), StoreError> {
+        let last_sent_height = self
+            .last_sent_height
+            .expect("syncer should have initialised the height by now");
+
+        let Some(lowest_range_height) = range.first().map(|h| h.height().value()) else {
+            // Ignore empty range
+            return Ok(());
+        };
+
+        if lowest_range_height < last_sent_height {
+            // We know range cannot cross last_sent_height, so either both ends are before or after.
+            // Ignore node syncing historical header ranges.
+            return self.inner.insert(range).await;
+        }
+
+        self.inner.insert(range.clone()).await?;
+
+        if last_sent_height + 1 == lowest_range_height {
+            self.send_range(range).await;
+        } else {
+            self.pending.push(range);
+        }
+
+        let mut i = 0;
+        while i < self.pending.len() {
+            if self.pending[i]
+                .first()
+                .expect("header range shouldn't be empty")
+                .height()
+                .value()
+                == self
+                    .last_sent_height
+                    .expect("last_sent_height should be initialised here")
+                    + 1
+            {
+                let range = self.pending.swap_remove(i);
```

Comment:
I prefer a `VecDeque` instead of this. but up to you

Reply:
Actually I changed my mind, this is gonna be nicer with deque

Reply:
Actually I changed my mind, this is gonna be nicer with deque

Reply:
nah, I was thinking we could order the ranges at insertion time with vecdeque push_front / push_back, but it doesn't work in case we loose connection, reconnect, don't have a time to sync the gap and loose connection again. Unless we insert and sort, vecdeque has no gains